### PR TITLE
Initial commit 1

### PR DESCRIPTION
**Pull request by camaeel**

### CODE REVIEW

**Comment by camaeel** [comment]
Golang implementation of task 1

**Comment by torgiren** [comment]
When staring bot on "empty" discord server, the bot is creating autorole channel and then failing with nul pointer
```
[torgiren@smartraptor torgiren-bot (issue-1)]$ docker run -e TOKEN=$TOKEN --rm -v $(pwd)/example/config.yaml:/config.yaml -it test 
INFO[0000] Bot is now running.  Press CTRL-C to exit.   
INFO[0000] Setting up guild: 997214659654463528         
INFO[0001] Channel autorole-channel created for guild 997214659654463528 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x6cc375]

goroutine 11 [running]:
main.setupAutorole(0xc0000b6000?, {0xc0004cc060, 0x4, 0x2?}, {0xc0000d6690, 0x12}, 0xc0000c06c0)
	/app/setup.go:46 +0x395
main.guildSetup(0x580?, 0xc00016a000, 0xc0000c0440)
	/app/setup.go:19 +0x189
main.main.func1(0x0?, 0x0?)
	/app/event_handlers.go:11 +0x25
github.com/bwmarrin/discordgo.guildCreateEventHandler.Handle(0x6baf0e?, 0xc00009f500?, {0x70e4e0?, 0xc000010018?})
	/go/pkg/mod/github.com/bwmarrin/discordgo@v0.25.0/eventhandlers.go:254 +0x38
created by github.com/bwmarrin/discordgo.(*Session).handle
	/go/pkg/mod/github.com/bwmarrin/discordgo@v0.25.0/event.go:171 +0x16a
```
possible defect:
- checking if channel exists
- if not, then creating a new one
- not getting new reference
- try to write a message to channel but using nil pointer

File: event_handlers.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		err = session.GuildMemberRoleAdd(event.GuildID, event.UserID, role.ID)
+		if err != nil {
+			log.Errorf("Couldn't add role %s to user %s in guild %s: %s", role.ID, event.UserID, event.GuildID, err)
+		} else {
+			log.Infof("Role %s added succesfully to user %s in guild %s", role.ID, event.UserID, event.GuildID)
+		}
+	}
+}
+
+func getRemoveReactionHandler(config *Config) func(*discordgo.Session, *discordgo.MessageReactionRemove) {
```
Comment (torgiren) [inline event_handlers.go:52]: try to deduplicate getRemoveReactionHandler and getAddReactionHandler functions

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,18 @@
+all: clean lint build test test-cover
+
+build: 
+	go build -o bin/discord-bot .
```
[the remark this answers is not in the document]
Comment (camaeel) [reply Makefile:4]: Added

File: setup.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	if err == nil {
+		log.Infof("Guild %s set up successfully", guild.ID)
+	} else {
+		log.Errorf("Guild %s setup failed", guild.ID)
+	}
+}
+
+func setupAutorole(session *discordgo.Session, channels []*discordgo.Channel, guildId string, autorole *Autorole) error {
```
[the remark this answers is not in the document]
Comment (camaeel) [reply setup.go]: Exactly. Fixed

File: setup.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	// find or crete channel
+	channel := FindChannel(channels, autorole.Channel)
+	if channel == nil {
+		channel, err := CreateChannel(session, guildId, autorole.Channel)
```
[the remark this answers is not in the document]
Comment (camaeel) [reply setup.go]: This was the case, channel has been assigned to a local variable, and never used

File: event_handlers.go (hunk 1 above)
Comment (camaeel) [reply event_handlers.go:39]: Improved a bit this one. But the result is IMHO not great.
I tried even using generics and/or closures.

**Comment by camaeel** [comment]
Closes #1

**Pull request merged**